consensus/src/round_manager.rs

```diff
@@ -531,7 +544,48 @@ impl RoundManager {
         );
 
         observe_block(proposal.timestamp_usecs(), BlockStage::SYNCED);
+        if self.decoupled_execution() && self.block_store.back_pressure() {
+            // In case of back pressure, we delay processing proposal. This is done by resending the
+            // same proposal to self after some time.
+            Ok(self
+                .resend_verified_proposal_to_self(
+                    proposal,
+                    BACK_PRESSURE_POLLING_INTERVAL_MS,
+                    BACK_PRESSURE_PROPOSAL_TIMEOUT_MS,
+                )
+                .await)
+        } else {
+            self.process_verified_proposal(proposal).await
+        }
+    }
 
+    async fn resend_verified_proposal_to_self(
+        &self,
+        proposal: Block,
+        polling_interval_ms: u64,
+        timeout_ms: u64,
+    ) {
+        let start = Instant::now();
+        let author = self.network.author();
+        let block_store = self.block_store.clone();
+        let self_sender = self.round_manager_tx.clone();
+        let event = VerifiedEvent::VerifiedProposalMsg(Box::new(proposal));
+        tokio::spawn(async move {
+            while start.elapsed() < Duration::from_millis(timeout_ms) {
+                if !block_store.back_pressure() {
+                    if let Err(e) =
+                        self_sender.push((author, discriminant(&event)), (author, event))
+                    {
+                        error!("Failed to send event to round manager {:?}", e);
+                    }
+                    break;
+                }
+                sleep(Duration::from_millis(polling_interval_ms)).await;
+            }
```

Comment:
is this all thread safe? (now that it is called from multiple threads)

Reply:
Yes, the `block_store` is designed to be accessed from multiple threads and the sender for the channel is thread safe as well.

Reply:
`block_store` can be read by multiple threads but probably not write